File: BOJ(Java 공부용)/1197.java
import java.io.*;
import java.util.PriorityQueue;
public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] input = br.readLine().split(" ");
        int v = Integer.parseInt(input[0]);
        int e = Integer.parseInt(input[1]);
        int[] parent = new int[v+1];
        PriorityQueue<p> pq = new PriorityQueue<>((a, b) -> {
            if(a.cost < b.cost) return -1;
            else if(a.cost > b.cost) return 1;
            else {
                if(a.here < b.here) return -1;
                else if(a.here > b.here) return 1;
                else {
                    return 0;
                }
            }
        });
        for(int i=1;i<=v;i++) parent[i] = i;
        for(int i=0;i<e;i++) {
            String[] input2 = br.readLine().split(" ");
            int here = Integer.parseInt(input2[0]);
            int next = Integer.parseInt(input2[1]);
            int cost = Integer.parseInt(input2[2]);
            pq.add(new p(here,next,cost));
        }
        int ans = 0;
        while(!pq.isEmpty()) {
            int here = pq.peek().here;
            int next = pq.peek().next;
            int cost = pq.peek().cost;
            pq.poll();
            if(find(here,parent)!= find(next,parent)) {
                union(here,next,parent);
                ans += cost;
            }
        }
        System.out.println(ans);
    }
    private static class p {
        int here,next,cost;
        private p(int here,int next,int cost) {
            this.here = here;
            this.next = next;
            this.cost = cost;
        }
    }
    private static int find(int u,int[] parent) {
        if(u == parent[u]) return u;
        return parent[u] = find(parent[u],parent);
    }
    private static void union(int u,int v,int[] parent) {
        u = find(u,parent);
        v = find(v,parent);
        if(u == v) return;
        parent[u] = v;
    }
}

File: BOJ(Kotlin)/1197.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

val bw = BufferedWriter(OutputStreamWriter(System.out))
fun main(args : Array<String>) {
    val br = BufferedReader(InputStreamReader(System.`in`))

    val itemList = br.readLine().split(" ")
    val v = itemList[0].toInt()
    val e = itemList[1].toInt()
    val parent = IntArray(v+1)
    val pq = PriorityQueue<P> { a, b -> a.cost - b.cost }

    for (i in 1..v) {
        parent[i] = i
    }
    for (i in 0 until e) {
        val inputList = br.readLine().split(" ")
        val here = inputList[0].toInt()
        val next = inputList[1].toInt()
        val cost = inputList[2].toInt()
        pq.add(P(here,next,cost))
    }

    var ans = 0
    while(!pq.isEmpty()) {
        val here = pq.peek().here
        val next = pq.peek().next
        val cost = pq.peek().cost
        pq.poll()
        if (find(here,parent) != find(next,parent)) {
            union(here,next,parent)
            ans += cost
        }
    }
    bw.write(ans.toString())
    bw.flush()
    bw.close()

}

fun find(u : Int, parent : IntArray) : Int{
    if (u == parent[u]) return u
    parent[u] = find(parent[u],parent)
    return parent[u]
}

fun union(u : Int, v : Int, parent : IntArray) {
    val u = find(u,parent)
    val v = find(v,parent)
    if (u == v) return
    parent[u] = v
}

class P(val here : Int, val next : Int, val cost : Int){

}